### PR TITLE
Tests: explicitly pass a ld for Darwin targets

### PR DESCRIPTION
**Pull request by compnerd**
This is based on #946 and transitively on #940.  Only the last commit is relevant.

### CODE REVIEW

**Comment by compnerd** [comment]
https://github.com/apple/swift-driver/pull/947/commits/047a7f0f7aa1a72eb8efe63ae51adac6f8dc7dd3?w=1 is helpful for viewing the diff.

**Review by cltnschlosser — APPROVED** [review]
Is specifying Darwin targets even valid on windows (or linux)?

This seems fine for now, since there is probably coverage that we wouldn't otherwise get, but I wonder if we need different test cases for other operating systems.

File: Tests/SwiftDriverTests/SwiftDriverTests.swift [hunk 1]
```diff
@@ -2987,22 +3036,44 @@ final class SwiftDriverTests: XCTestCase {
...
-    XCTAssert(plannedJobs[0].commandLine.contains(.flag("-target")))
-    XCTAssertFalse(plannedJobs[0].commandLine.contains(.flag("-clang-target")))
+    try withTemporaryDirectory { tmpDir in
+      // Provide an override for ld as not all platforms have a `ld`
+      let ld = tmpDir.appending(component: executableName("ld64.lld"))
+      try localFileSystem.writeFileContents(ld) { $0 <<< "" }
+      try localFileSystem.chmod(.executable, path: AbsolutePath(ld.pathString.nativePathString()))
+
+      var envVars = ProcessEnv.vars
+      envVars["SWIFT_DRIVER_LD_EXEC"] = ld.pathString.nativePathString()
```
[the remark this answers is not in the document]
Comment (compnerd) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift:3046]: Interesting thought; I'd say that is a good follow up change to do.

**Comment by compnerd** [comment]
> Is specifying Darwin targets even valid on windows (or linux)?

Yes and no.  Yes, that is entirely valid thing to do, no you do not expect it to get anywhere.  The reason for the latter is that you do not have the platform SDK.  That is only available through Xcode.  But, assuming that you had the SDK, then this is totally valid to do this on other platforms.

> I wonder if we need different test cases for other operating systems.

Yes, in general, we should prefer to use the host target and only explicitly specify the target triple when we are testing target specific behaviour rather than general driver behaviour.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

File: Tests/SwiftDriverTests/SwiftDriverTests.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (compnerd) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift]: I change my mind; I've done it in this change as a helper `withMockLinker`.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Review by artemcm — APPROVED** [review]

**Pull request merged**